### PR TITLE
[#8785] Remove some interdependencies in twisted.mail

### PR DESCRIPTION
**Pull request by hawkowl**
This should make it easier to port, and make those GIANT files much nicer to deal with.

### CODE REVIEW

File: src/twisted/mail/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,1070 @@
...
+    An interface for email domains.
+    """
+    def exists(user):
+        """
+        Check whether a user exists in this domain.
+
+        @type user: L{User}
+        @param user: A user.
+
+        @rtype: no-argument callable which returns L{IMessageSMTPI} provider
```
Comment (rodrigc) [inline src/twisted/mail/interfaces.py]: Typo here is causing a documentation build error.

File: src/twisted/mail/interfaces.py [hunk 2]
```diff
@@ -0,0 +1,1070 @@
...
+        @type flags: sequence of C{str}
+        @param flags: The flags to set, unset, or add.
+
+        @type mode: -1, 0, or 1
+        @param mode: If mode is -1, these flags should be removed from the
+            specified messages. If mode is 1, these flags should be added to
+            the specified messages. If mode is 0, all existing flags should be
+            cleared and these flags should be added.
+
+        @type uid: C{bool}
```
Comment (rodrigc) [inline src/twisted/mail/interfaces.py]: Take a pass through the docstring for what you have changed and change:

C{bool} -> L{bool}
C{str} -> L{str}
C{dict} -> L{dict}
C{tuple} -> L{tuple}
C{int} -> L{int}
C{Deferred} -> L{Deferred}

etc.

**Comment by rodrigc** [comment]
You need a topfile as well.  Please address my review comments.

**Comment by hawkowl** [comment]
This is still waaaay too big

**Comment by hawkowl** [comment]
See https://github.com/twisted/twisted/pull/538

**Comment by rodrigc** [comment]
More moving stuff around.  OK to merge.

**Pull request merged**